python-oauth-web-app/app.py

```diff
@@ -0,0 +1,300 @@
+        {
+            # simply using the full request URL for the client_id
+            "client_id": client_id,
+            "dpop_bound_access_tokens": True,
+            "application_type": "web",
+            "redirect_uris": [f"{app_url}oauth/authorize"],
+            "client_uri": app_url,
+            "subject_type": "public",
+            "grant_types": ["authorization_code", "refresh_token"],
+            "response_types": ["code"],  # TODO: "code id_token"?
+            "scope": "openid profile offline_access",
+            "client_name": "atproto OAuth Flask Backend Demo",
+            "token_endpoint_auth_method": "none",
+            "jwks_uri": f"{app_url}oauth/jwks.json",
+            # "jwks": { #    "keys": [pub_jwk], #},
+        }
+    )
+
+
+@app.route("/oauth/jwks.json")
+def oauth_jwks():
+    return jsonify(
+        {
+            "keys": [pub_jwk],
+        }
+    )
+
+
+@app.route("/oauth/login", methods=("GET", "POST"))
+def oauth_login():
+    if request.method != "POST":
+        return render_template("login.html")
+
+    # resolve handle and DID document, using atproto identity helpers
+    username = request.form["username"]
+    if not valid_handle(username):
+        flash("Invalid Handle")
+        return render_template("login.html"), 400
+    did = resolve_handle(username)
+    if not did:
+        flash("Handle Not Found")
+        return render_template("login.html"), 400
+    if not valid_did(did):
+        flash("Handle resolved to invalid DID")
+        return render_template("login.html"), 400
+    did_doc = resolve_did(did)
+    if not did_doc:
+        flash("DID Not Found: " + did)
+        return render_template("login.html"), 400
+
+    # fetch PDS OAuth metadata
+    # IMPORTANT: PDS endpoint URL is untrusted input, SSRF mitigations are needed
+    pds_url = pds_endpoint(did_doc)
+    print(f"PDS: {pds_url}")
+    assert is_safe_url(pds_url)
+    with hardened_http.get_session() as sess:
+        resp = sess.get(f"{pds_url}/.well-known/oauth-protected-resource")
+    resp.raise_for_status()
+
+    # IMPORTANT: Authorization Server URL is untrusted input, SSRF mitigations are needed
+    authsrv_url = resp.json()["authorization_servers"][0]
+    print(f"Auth Server: {authsrv_url}")
+    assert is_safe_url(authsrv_url)
+    try:
+        authsrv_meta = fetch_authsrv_meta(authsrv_url)
+    except Exception as err:
+        print(f"failed to fetch authsrv_meta: " + str(err))
+        flash("Failed to fetch Auth Server (Entryway) OAuth metadata")
+        return render_template("login.html"), 400
+
+    app_url = request.url_root.replace("http://", "https://")
+    client_id = f"{app_url}oauth/client-metadata.json"
+
+    pkce_verifier, state, nonce, resp = send_par_auth_request(
+        authsrv_url, authsrv_meta, did, app_url, secret_jwk
+    )
+    resp.raise_for_status()
+    request_uri = resp.json()["request_uri"]
+
+    # create a DPoP keypair for this user session now (TODO: why now?)
```

Comment:
Only useful to generate it now if you use it as dpop proof for the PAR request, which will bind the authorization request to that dpop key.

With the current implementation here, there is indeed no point of generating it now.

I would advise to provide a dpop proof as part of the send_par_auth_request.

Reply:
For some reason, the PAR + DPoP spec allow you to either use a dpop key from the very beginning (during the PAR), or when you obtain the tokens.

The atproto spec does not enforce anything on top, though we might (?).

Reply:
updated to have DPoP start at PAR